draft-ietf-dprive-dnsoquic.md

```diff
@@ -552,31 +552,27 @@ connections. For example:
 
 ## Flow Control Mechanisms
 
-Servers and Clients manage flow control as specified in QUIC.
-
-Servers can use the "maximum stream ID" parameter of the QUIC transport to limit
-the number of streams opened by the client. This mechanism will effectively
-limit the number of DNS queries that a client can send on a single DoQ
-connection. The initial value of this parameter is specified by the transport
-parameter `initial_max_streams_bidi`. To allow at least one query per client,
-servers SHOULD use a non-zero value for this parameter. Using larger values
-will result in better performance.
-
-The flow control mechanisms of QUIC control how much data can be sent by QUIC
-nodes at a given time. The initial values of per stream flow control parameters
-is defined by two transport parameters:
-
-* initial_max_stream_data_bidi_local: when set by the client, specifies the
-  amount of data that servers can send on a "response" stream without waiting
-  for a MAX_STREAM_DATA frame.
-
-* initial_max_stream_data_bidi_remote: when set by the server, specifies the
-  amount of data that clients can send on a "query" stream without waiting for
-  a MAX_STREAM_DATA frame.
-
-For better performance, it is RECOMMENDED that clients and servers set each of
-these two parameters to a value of 65537 or greater, allowing for at least one
-query or response of maximum length 65535 bytes.
+Servers and Clients manage flow control using the mechanisms defined in
+section 4 of {{!RFC9000}}. These mechanisms allows clients and servers to
+how many streams can be created, how much data can be sent on a stream,
+and how much data can be sent on the union of all streams. For DNS over QUIC,
+controlling how many streams are created allow servers to control how many
+new requests the client can send on a given connection.
+
+The global and per stream flow control allow servers to control how much data can be sent by
+clients. If that control is strict, clients will have to wait for new credits
+before they can transmit the full content of the request. The same mechanisms
+allow clients to control how much data can be sent by servers. If too strict,
+servers will have to wait for credits before sending the full content of the
+responses.
+
+Initial values of parameters control how many requests and how much data can be
+sent by clients and servers at the beginning of the connection. These values
+are specified in transport parameters exchanged during the connection handshake.
+The parameter values received in the initial connection also control how many requests and
+how much data can be sent by clients using 0-RTT data in a resumed connection.
+Using too small values of these initial parameters would restrict the
+usefulness of allowing 0-RTT data.
```

Comment:
Generic advice about the trade-off might be clearer than specific advice...

```suggestion

Flow control exists to protect endpoint resources.
Values that are too small will unnecessarily limit performance.
Values that are too large might expose endpoints to overload or memory exhaustion.
Implementations or deployments will need to adjust flow control limits to balance these concerns.
```

This is duplicative of the sentence on line 565-567, which might be removed if you like this text.

Reply:
There is another issue there. Flow control credits can be given through initial values (transport parameters) or by using dynamic updates (various flow control frames), but 0-RTT data is solely controlled by the initial values. Since 0-RTT performance is important, I want to point that out.

Reply:
Let's try a compromise...

Reply:
I wonder if we should add some specific advice here around XFR because use of values that are too small in that case could prevent a large IXFR or a full AXFR being sent (not just a performance issue).... this *could* also be true for large DNSSEC response....

Reply:
@saradickinson there are two issues: the initial value of the parameters, and the ongoing value. For 0-RTT, flow control is entirely defined by the initial value. That limits the amount of data that clients can send. But clients should only send queries. Even with EDNS options, queries are rather short, so I do not expect much issue there.

In the case of AXFR, unless the initial limit is huge, the client will most probably need to update the flow control limits during the course of the connection. That's standard behavior in QUIC. For example, in tests of 1GB downloads, the initial limit is typically in the 64KB to 1MB range, and then the client gradually increases it -- how "gradually" being left to implementers.

Martin's point is that we should try not to duplicate or paraphrase the QUIC standard in this document. Indeed, doing so can backfire if there is a slight contradiction between what we write and the QUIC RFCs. I wanted to point out the effect on 0-RTT, because that's somewhat specific to the DoQ application. But the handling of large XFR is not fundamentally different from downloading large documents with HTTP, so I don't think that we need extra text.

On the other hand, we may need something else, a reference in the security section to the Slowloris attack (section 21.6 or RFC 9000). That attack can be instantiated in DoQ just like in H3. Clients could send a bunch of queries, or start AXFR, and then play games with flow control credits to stall the server. Maybe we should mention that.